### PR TITLE
Cutout Tools

### PR DESCRIPTION
**Pull request by robelgeda**
This PR introduces cutout tools used to extract postage stamp images of multiple targets. The functions take in images (multiple formats) and a catalog of targets to produce cutouts. The motivation behind this PR is the increased demand for producing cutouts from catalogs. For example, after conducting multi slit spectroscopy, the user wants to produce cutouts of each spectral targets from a field image. 

There are two functions introduced:
`make_cutouts` :  Makes cutouts of catalog targets from a 2D image array. 

`cutouts_from_fits` : Wrapper for the make_cutouts function. This function will take in a single
    fits image and return an array containing a list of cutouts as fits hdus. It
    will also save the cutouts to file if requested.

### CODE REVIEW

**Comment by pllim** [comment]
Might be too late for 3.1? If not, feel free to re-milestone.

File: .travis.yml [hunk 1]
```diff
@@ -33,7 +33,7 @@ env:
         - NUMPY_VERSION=stable
         - MAIN_CMD='python setup.py'
         - CONDA_DEPENDENCIES='Cython jinja2'
-        - CONDA_ALL_DEPENDENCIES='Cython jinja2 scipy h5py matplotlib pyyaml scikit-image pandas pytz beautifulsoup4 ipython mpmath bleach asdf bottleneck'
+        - CONDA_ALL_DEPENDENCIES='Cython jinja2 scipy h5py matplotlib pyyaml scikit-image pandas pytz beautifulsoup4 ipython mpmath bleach asdf bottleneck reproject'
```
[the remark this answers is not in the document]
Comment (bsipocz) [reply .travis.yml]: hmm, using an affiliated package that also depends on astropy? It'll be a bit interesting how conda handles it (I suspect it'll install the stable astropy, and then I'm not sure what happens...).

[the remark this answers is not in the document]
Comment (bsipocz) [reply .travis.yml]: However this makes me wonder, do we have an ETA for reproject landing in core? If it's fairly soon, e.f. 3.2, wouldn't it be possible to add it into extern for this 3.1 release?

Or add this feature to reproject instead until it lands in astropy core?

File: astropy/nddata/utils.py [hunk 2]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
         are for the valid (non-filled) cutout values.
         """
         return self._calc_bbox(self.slices_cutout)
+
+
+def make_cutouts(data, catalog, wcs=None, origin=0, verbose=True):
+    """Make cutouts of catalog targets from a 2D image array.
+
+    Notes
```
Comment (larrybradley) [inline astropy/nddata/utils.py]: The `Notes` section should go below the `Returns` section (we follow the numpydoc docstring style).

File: astropy/nddata/utils.py [hunk 3]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
...
+    -----
+    The input Catalog must have the following columns, which must have
+    `~astropy.unit.Unit`s where applicable:
```
Comment (larrybradley) [inline astropy/nddata/utils.py]: A `QTable` with `Quantity` mixin columns should also work.

File: astropy/nddata/utils.py [hunk 4]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
...
+        * ``OBJ_DEC`` - DEC of the cutout object in degrees.
+        * ``OBJ_ROT`` - Rotation of cutout object in degrees.
+
+    Parameters
+    ----------
+    data : 2D `~numpy.ndarray` or `~astropy.nddata.NDData`
+        The 2D cutout array.
+    catalog : `~astropy.table.table.Table`
+        Catalog table defining the sources to cut out. Must contain
+        unit information as the cutouttool does not assume default units.
```
Comment (larrybradley) [inline astropy/nddata/utils.py]: cutouttool -> cutout tool

File: astropy/nddata/utils.py [hunk 5]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
...
+    wcs : `~astropy.wcs.wcs.WCS`
+        WCS if the input image is `~numpy.ndarray`.
+    origin : int
+        Whether SkyCoord.from_pixel should use 0 or 1-basedpixel coordinates.
```
Comment (larrybradley) [inline astropy/nddata/utils.py]: basedpixel -> based pixel

File: astropy/nddata/utils.py [hunk 6]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
...
+    -------
+    cutouts : list
+        A list of NDData. If cutout failed for a target,
+       `None` will be added as a place holder.
+    """
+    # Optional dependencies...
+    try:
+        from reproject.interpolation.high_level import reproject_interp
+    except ImportError as e:
+        raise ImportError("Optional requirement not met: " + e.msg)
```
Comment (larrybradley) [inline astropy/nddata/utils.py]: I would probably check for this dependency only if `reproject` is actually needed (i.e. if rotation was requested; `apply_rotation=True`)

File: astropy/nddata/utils.py [hunk 7]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
...
+
+    cutcls = partial(Cutout2D, data.data, wcs=wcs, mode='partial')
+    cutouts = []
+    for position, x_pix, y_pix, row in zip(coords, width, height, catalog):
+        if apply_rotation:
+            pix_rot = row['cutout_pa'].to(u.degree).value
+
+            # Construct new rotated WCS:
+            cutout_wcs = WCS(naxis=2)
+            cutout_wcs.wcs.ctype = ['RA---TAN', 'DEC--TAN']
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply astropy/nddata/utils.py]: And this needs to be clearly stated in the doc string.

File: astropy/nddata/utils.py [hunk 8]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
...
+    return cutouts
+
+
+def cutouts_from_fits(image, catalog, image_ext=0, origin=0,
+                      output_dir=None, overwrite=False, verbose=True):
+    """Wrapper for the make_cutouts function. This function will take in a single
+    fits image and return an array containing a list of cutouts as fits hdus. It
+    will also save the cutouts to file if requested.
+
+    Notes
```
Comment (larrybradley) [inline astropy/nddata/utils.py]: The `Notes` section should go after the `Returns` section.

File: astropy/nddata/utils.py [hunk 9]
```diff
@@ -916,3 +925,373 @@ def bbox_cutout(self):
...
+            <id>.fits
+
+    Each cutout image is a simple single-extension FITS with updated WCS.
+    Its header has the following special keywords:
+
+        * ``OBJ_RA`` - RA of the cutout object in degrees.
+        * ``OBJ_DEC`` - DEC of the cutout object in degrees.
+        * ``OBJ_ROT`` - Rotation of cutout object in degrees.
+
+    Examples
```
Comment (larrybradley) [inline astropy/nddata/utils.py]: The `Examples` section goes after the `Notes` section (at the bottom).  `Examples` are always last.

**Comment by astrofrog** [comment]
I plan to review this in detail tomorrow, but regarding the reproject dependency, one option if we want to get this into v3.1 is to remove the ability to have rotated cutouts for now and then figure out during the v3.2 development cycle how we want to handle that.

File: astropy/nddata/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (robelgeda) [reply astropy/nddata/utils.py]: I added a check for the input WCS for now.

File: astropy/nddata/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (robelgeda) [reply astropy/nddata/utils.py]: I am not sure what you mean hear, can you please give an example?

**Comment by robelgeda** [comment]
@eteq @larrybradley : I have made the edits as requested.

**Comment by astrofrog** [comment]
I just talked with @eteq about this, and we have a suggested plan of action :)

This PR actually is doing two things: making it possible to make rotated cutouts, and providing a convenience to go from a table to cutouts, and @eteq has convinced me that both are worthwhile. However, we agreed that the issue is that the rotation needs to also be available to people who want to use ``Cutout2D`` directly - that is, the convenience function shouldn't have cutout functionality not present in ``Cutout2D``.

So our suggestion is that this PR should be split into two:

* One PR that adds the rotation functionality to ``Cutout2D``
* One PR that adds a convenience function to go from an image and a table to cutouts

And this could maybe even be done in three steps:

* One PR that adds a convenience function to go from an image and a table to cutouts, **without rotation**
* One PR that adds the rotation functionality to ``Cutout2D``
* One PR that adds the ability to specify rotation in the convenience function

Splitting things into smaller PRs is actually a good recipe for success, as it's often otherwise difficult to reach consensus.

While it seems unlikely that some of this can be done before the v3.1 feature freeze, we should keep up the momentum and still get all this done as soon as possible, even if it's after feature freeze. I know that this is needed for MOSViz, but it doesn't have to be in v3.1 for that - as soon as it's merged into the core package, you can copy a version back to MOSViz and bundle it there, and then do something like:

```python
try:
    from astropy.nddata.utils import make_cutouts_from_table
except ImportError:
    from mosviz.compat.nddata.utils import make_cutouts_from_table
```

That is, for a limited time you can include a bundled version in MOSViz, but this will be forward-compatible with Astropy v3.2. I've done this frequently with other functionality I needed to rely on from the core astropy package that wasn't in a released version yet.

**Comment by larrybradley** [comment]
There is a lot of functionality in `Cutout2D` to map cutout <--> origin image positions, slices, bounding boxes, etc. via its attributes.  None of those features will be applicable to a rotated cutout since cutout pixels will no longer map 1-to-1 to pixels in the original image.  Because of this, I don't think rotation should be added to `Cutout2D`.  It should probably go into a separate class, e.g. `RotatedCutout2D`.  In that case, the basic array cutout functionality should probably go into a common base class and then `Cutout2D` could be a subclass that adds all of the cutout <--> original image mapping, and `RotatedCutout2D` could be a subclass that adds the rotation (but not the mapping stuff).  However, I'm not in favor of rushing this through before the v3.1 feature freeze.

**Comment by robelgeda** [comment]
I really like the discussion here and also think its best to break this PR down. I also do not think we should rush it (I think the deadline is tomorrow).

**Comment by bsipocz** [comment]
Given the consensus above I've remilestoned this.

**Comment by mcara** [comment]
What is the relationship (if any) between this Cutout and [`Cutout2D`](https://github.com/astropy/astropy/blob/d0e840de832e44aa77ecec871f06042535bd0821/astropy/nddata/utils.py#L457)?

**Comment by mcara** [comment]
After reading this a little bit more carefully, I agree with @astrofrog comment https://github.com/astropy/astropy/pull/7940#pullrequestreview-168482498

In addition, I feel the way rotations are handled may not be robust enough. I think `reproject` itself should be able to handle rotations given an input WCS and a rotated WCS.

**Comment by mwcraig** [comment]
Moving the milestone to 4.0

**Pull request closed without merging**